### PR TITLE
Lesson12

### PR DESCRIPTION
**Pull request by vtx123**

### CODE REVIEW

File: Lesson12/Lesson12_1.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                arr[i + 1] = tmp;
+            }
+        }
+    }
+
+    std::cout << "Bubble sorted array: "<< std::endl;
+    for(int i = 0; i < size; i++){
+        std::cout << "[" << arr[i] << "]";
+    }
+    std::cout << std::endl;
```
Comment (helen-black) [inline Lesson12/Lesson12_1.cpp]: you should print it outside this function

File: Lesson12/Lesson12_2.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            for(int j = 0; j < ROWS; j++) {
+                arrByRow[j] = arr[j][i];
+            }
+            quickSort(arrByRow, 0, ROWS);
+
+            for(int k = 0; k < ROWS; k++) {
+                arr[k][i] = arrByRow[k];
+            }
+        }
+        std::cout << "Sorted by rows array: " << std::endl;
```
Comment (helen-black) [inline Lesson12/Lesson12_2.cpp]: functions shouldn't talk. move it outside

File: Lesson12/Lesson12_2.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    } else {
+        for(int i = 0; i < ROWS; i++){
+            for(int j = 0; j < COLUMNS; j++) {
+                arrByCol[j] = arr[i][j];
+            }
+            quickSort(arrByCol, 0, COLUMNS);
```
Comment (helen-black) [inline Lesson12/Lesson12_2.cpp]: you don't need additional array here. you can call sort like this:
`quickSort(arr[i], 0, COLUMNS);`

File: Lesson12/Lesson12_2.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        quickSort(array, low, partition_idx - 1);
+        quickSort(array, partition_idx + 1, high);
+    }
+}
+
+void sort(int arr[ROWS][COLUMNS], SortingDirection sd)
+{
+    int arrByRow[ROWS];
+    int arrByCol[COLUMNS];
+    if(sd == SortingDirection::ByRows){
```
Comment (helen-black) [inline Lesson12/Lesson12_2.cpp]: you mixed up sort byRows and byColumns. sort by row is when your row index static and you change col index, so it is row: `a[0][0], a[0][1], a[0][2]`, and column is `a[0][0], a[1][0], a[2][0]`

File: Lesson12/Lesson12_3.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    return false;
+}
+
+// ??? Evaluate algorithmic complexity of this program
+
+// ANSWER: binarySearch function will run 1-20 times.
+// Code inside binarySearch function will run:
+// Best case = 1 * 1 = 1 times = O(1)
+// Worst case = 20 * 5 = 100 times = O(r * log(c)), r = ROWS, c = COLUMNS.
```
Comment (helen-black) [inline Lesson12/Lesson12_3.cpp:32]: you forgot about space complexity

**Review by helen-black — CHANGES_REQUESTED** [review]

File: Lesson12/Lesson12_1.cpp (hunk 1 above)
Comment (vtx123) [reply Lesson12/Lesson12_1.cpp]: fixed

File: Lesson12/Lesson12_2.cpp (hunk 2 above)
Comment (vtx123) [reply Lesson12/Lesson12_2.cpp]: fixed

File: Lesson12/Lesson12_2.cpp (hunk 3 above)
Comment (vtx123) [reply Lesson12/Lesson12_2.cpp]: fixed

File: Lesson12/Lesson12_2.cpp (hunk 4 above)
Comment (vtx123) [reply Lesson12/Lesson12_2.cpp]: looks like fixed

File: Lesson12/Lesson12_3.cpp (hunk 5 above)
Comment (vtx123) [reply Lesson12/Lesson12_3.cpp:32]: added

**Review by helen-black — APPROVED** [review]

**Pull request merged**